crates/ethereum-rpc/src/lib.rs

```diff
@@ -638,52 +565,135 @@ fn register_rpc_methods<C: sov_modules_api::Context, Da: DaService>(
-            // also since this is not cached we need to stop at somewhere, so we add param stop_at
-            if opts.as_ref().map_or(true, |o| o.tracer.is_none()) {
-                let traces = evm.trace_block_transactions_by_number(
-                    block_number,
-                    opts.clone(),
-                    Some(trace_index as usize),
-                    &mut working_set,
-                )?;
-                return Ok(traces[trace_index as usize].clone());
-            }
-
-            // check cache if found convert to requested tracer and config and return
-            if let Some(traces) = ethereum.trace_cache.lock().unwrap().get(&block_number) {
-                let requested_opts = opts.clone().unwrap();
-                let tracer_config = requested_opts.tracer_config;
-                let traces = get_traces_with_reuqested_tracer_and_config(
-                    vec![traces[trace_index as usize].clone()],
-                    requested_opts.tracer.unwrap(),
-                    tracer_config,
-                )?;
-                return Ok(traces.into_iter().next().unwrap());
-            }
-
-            let cache_options = create_trace_cache_opts();
-            let traces = evm.trace_block_transactions_by_number(
+            let traces = debug_trace_by_block_number(
                 block_number,
-                Some(cache_options),
-                None,
+                Some(trace_idx as usize),
+                &ethereum,
+                &evm,
                 &mut working_set,
+                opts,
             )?;
-            ethereum
-                .trace_cache
-                .lock()
-                .unwrap()
-                .insert(block_number, traces.clone());
-            // Convert the traces to the requested tracer and config
-            let requested_opts = opts.clone().unwrap();
-            let tracer_config = requested_opts.tracer_config;
-            let traces = get_traces_with_reuqested_tracer_and_config(
-                vec![traces[trace_index as usize].clone()],
-                requested_opts.tracer.unwrap(),
-                tracer_config,
-            )?;
+            Ok(traces[0].clone())
+        },
+    )?;
 
-            Ok(traces.into_iter().next().unwrap())
+    rpc.register_subscription(
+        "debug_subscribe",
+        "debug_subscription",
+        "debug_unsubscribe",
+        |parameters, pending, ethereum| async move {
+            let mut params = parameters.sequence();
+
+            let topic: String = match params.next() {
+                Ok(v) => v,
+                Err(err) => {
+                    pending.reject(err).await;
+                    return Ok(());
+                }
+            };
+            match topic.as_str() {
+                "traceChain" => {
+                    let start_block: BlockNumberOrTag = match params.next() {
+                        Ok(v) => v,
```

Comment:
move this match arm to some function like `handle_trace_chain`

Reply:
so that when we have another arm we respond to, code is more readable

Reply:
Done.